caffe2/core/blob.h

```diff
@@ -30,36 +30,23 @@ class CAFFE2_API Blob final {
   /**
    * Initializes an empty Blob.
    */
-  Blob() : meta_(), pointer_(nullptr) {}
+  Blob() noexcept : meta_(), pointer_(nullptr), destroy_(nullptr) {}
   ~Blob() { Reset(); }
 
-  Blob(Blob&& other) noexcept
-      : meta_(std::move(other.meta_)),
-        pointer_(std::move(other.pointer_)),
-        destroy_(std::move(other.destroy_)) {
-    other.meta_ = {};
-    other.pointer_ = nullptr;
-    other.destroy_ = nullptr;
+  Blob(Blob&& other) noexcept : Blob() {
+    swap(other);
   }
 
   Blob& operator=(Blob&& other) noexcept {
-    if (pointer_ && destroy_) {
-      destroy_(pointer_);
-    }
-    meta_ = std::move(other.meta_);
-    pointer_ = std::move(other.pointer_);
-    destroy_ = std::move(other.destroy_);
-    other.meta_ = {};
-    other.pointer_ = nullptr;
-    other.destroy_ = nullptr;
+    Blob(std::move(other)).swap(*this);
```

Comment:
This is going to marginally less efficient while TypeMeta is a giant object, because you're doing some extra, unnecessary moves on the old meta before killing it.

Reply:
I know. When I was doing the inverse of this change in another PR for IValue, @zdevito pointed out that it is advantageous if it doesn't need to know the implementation details of the class, and I agree. I actually don't really mind which pattern we follow, but I want them aligned.